src/compiler/checker.ts

```diff
@@ -14544,6 +14544,17 @@ namespace ts {
                 getIndexInfoWithReadonly(numberIndexInfo, readonly));
             spread.objectFlags |= ObjectFlags.ObjectLiteral | ObjectFlags.ContainsObjectOrArrayLiteral | ObjectFlags.ContainsSpread | objectFlags;
             return spread;
+
+            function errorTypeIfTooLarge(): Type | undefined {
+                if (left.flags & right.flags & TypeFlags.Union) {
+                    const resultSize = (left as UnionType).types.length * (right as UnionType).types.length;
+                    if (resultSize > 100000) {
+                        tracing.instant(tracing.Phase.Check, "getSpreadType_DepthLimit", { leftId: left.id, rightId: right.id });
+                        error(currentNode, Diagnostics.Spread_expression_produces_a_union_type_that_is_too_complex_to_represent);
+                        return errorType;

```

Comment:
@minestarks raised an interesting point: should this be `anyType` so that users have the option of ts-ignoring this error?